### PR TITLE
changing permission for kubeconfig file inside the shell pod

### PR DESCRIPTION
**Pull request by diogoasouza**
This PR changes the permission of the kubeconfig file used in the `shell` container used by the impersonator. It was using 0644 which was causing warnings when using helm. This changes the permission to 0600.

**Explanation:**
Kubernetes always [mounts](https://github.com/kubernetes/kubernetes/blob/46a2137c1ba017970c316c0ec10c074cb6450732/pkg/kubelet/kubelet_pods.go#L344) [configMaps](https://github.com/kubernetes/kubernetes/blob/46a2137c1ba017970c316c0ec10c074cb6450732/pkg/volume/configmap/configmap.go#L164) and `Secret` volumes as `readOnly`, regardless of the flag passed when creating the mount. This is [not reflected](https://github.com/kubernetes/api/blob/dd2e75089b339ee9b1b2aed2c275bf905079dde3/core/v1/types.go#L2155) in the documentation of the `readOnly` flag. To work around that, I'm using an `init-container` to copy the content to an `emptyDir` volume, changing the permission and using that volume in the `shell` container

### CODE REVIEW

File: pkg/podimpersonation/podimpersonation.go
Comment (rohitsakala) [inline pkg/podimpersonation/podimpersonation.go:584]: Should we use our own busybox ? rancher/mirrored-bci-busybox or rancher/library-busybox. I don't know which is the latest one. Maybe ask @pjbgf ?

File: pkg/podimpersonation/podimpersonation.go
Comment (rohitsakala) [inline pkg/podimpersonation/podimpersonation.go:545]: nit: user-kube-configmap to

File: pkg/podimpersonation/podimpersonation.go
Comment (rohitsakala) [inline pkg/podimpersonation/podimpersonation.go:599]: Do we need this subpath?

File: pkg/podimpersonation/podimpersonation.go
Comment (pjbgf) [reply pkg/podimpersonation/podimpersonation.go:584]: `registry.suse.com/bci/bci-busybox`.

File: pkg/podimpersonation/podimpersonation.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go:599]: We only use what's inside the `config` key of the ConfigMap. We don't _need_ it, but I think it's good to specify a `SubPath` if we know exactly what we need

File: pkg/podimpersonation/podimpersonation.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: Good call, done

File: pkg/podimpersonation/podimpersonation.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: Done

File: pkg/podimpersonation/podimpersonation.go
Comment (pjbgf) [inline pkg/podimpersonation/podimpersonation.go:519]: Using Go's octal syntax makes it slightly more readable:
```suggestion
		m         = 0o644
		m2        = 0o600
```

Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: Done

**Comment by diogoasouza** [comment]
Can I get a review from someone on the frameworks team?
@rancher/rancher-squad-frameworks

File: pkg/podimpersonation/podimpersonation.go
Comment (rohitsakala) [inline pkg/podimpersonation/podimpersonation.go:583]: Can we rename to something descriptive ? instead of `init-container`

File: pkg/podimpersonation/podimpersonation.go
Comment (rohitsakala) [inline pkg/podimpersonation/podimpersonation.go:582]: Due to the new `init-container` when I click on the `Kubectl shell` button on Rancher Manager, it is taking 18 seconds to load the shell.

Before the `init-container` it takes only 4 seconds to load the shell.

This is a serious degradation of user experience.

Maybe due to this, I find the `sed` solution better, if we can't find a way to mitigate this user experience issue.

**Comment by tomleb** [comment]
Can you link a GH issue associated with this change?

I'm also not very familiar with this area but I'll do my best at reviewing this.

 From what I understand, we want to mount some kubeconfig that's stored on a `ConfigMap` and we want it to have `0600` privileges to avoid a warning from the `helm` binary.

Is it not possible to mount this file with the `defaultMode` set to 0600? I see you're using it for the one in the init container, but why not in the container itself directly?

Here's an example (taken and modified from the upstream k8s docs):

```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: game-config
  namespace: default
data:
  game.properties: |
    enemies=aliens
    lives=3
    enemies.cheat=true
    enemies.cheat.level=noGoodRotten
    secret.code.passphrase=UUDDLRLRBABAS
    secret.code.allowed=true
    secret.code.lives=30    
  ui.properties: |
    color.good=purple
    color.bad=yellow
    allow.textmode=true
    how.nice.to.look=fairlyNice    
---
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: registry.k8s.io/busybox
      command: [ "/bin/sh", "-c", "ls -l /etc/config/" ]
      volumeMounts:
      - name: game-config
        subPath: game.properties
        mountPath: /etc/config/game.properties
  volumes:
    - name: game-config
      configMap:
        name: game-config
        defaultMode: 0600
  restartPolicy: Never
```

Looking at the logs of the pod:

```sh
$ kubectl logs dapi-test-pod
total 4
-rw-------    1 root     root           162 Mar  8 13:12 game.properties
```

The mode bits are as expected.

**Comment by diogoasouza** [comment]
@tomleb Hi Tom, thanks for the review. 
Regarding that, it's true that we can change the `defaultMode`, the problem is that the user that [runs](https://github.com/rancher/shell/blob/b88ba63c4a0da560fa99f9ca9a357023440c0b42/package/Dockerfile#L71) inside the `shell` container is not the `root` user. Setting the `runAsUser` in the container does not change the ownership of the files inside the mounted volume as well

File: pkg/podimpersonation/podimpersonation.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go:582]: It was the `imagePullPolicy` that was causing this. The default value is `Always` and I changed it to `PullIfNotPresent`. Should be fine now

File: pkg/podimpersonation/podimpersonation.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: Done

File: pkg/podimpersonation/podimpersonation.go
Comment (tomleb) [inline pkg/podimpersonation/podimpersonation.go:584]: I'm not a fan of using `:latest` tags for images. Something that used to work might stop working randomly if that `:latest` image breaks.

Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: Makes sense. Changed it to use `registry.suse.com/bci/bci-busybox:15.5.16.1` which is the latest one available as of today

Comment (pjbgf) [reply pkg/podimpersonation/podimpersonation.go]: +1 on not using `latest`. Overall we tend to recommend folks tag to the minor: `registry.suse.com/bci/bci-busybox:15.5`. So you latest patches, without the risk of breaking changes.

Comment (tomleb) [reply pkg/podimpersonation/podimpersonation.go]: Michael reminded me about airgap. It seems like hardcoding `registry.suse.com/bci/bci-busybox:whatever` would not work for airgap use cases where the customer does not have access to the SUSE registry.

@diogoasouza one idea here would be to re-use the shell image but change its entrypoint via `command:`. That would avoid making the busybox available for airgap use case (no idea if it's already done btw).

Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: @tomleb I checked and the busybox image is mirrored, but only the `15.4` version, and that was more than one year ago. I changed the init container to use the shell image, it shouldn't affect the latency since the image is going to have to be pulled anyway for the `proxy` and `shell` containers.

File: pkg/podimpersonation/podimpersonation.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go:600]: @rohitsakala Yes, sorry if I wasn't clear. I meant that we only copy the `config` from the `configMap` to this volume and that's why we only need to mount this path to the container

**Review by rohitsakala — APPROVED** [review]

File: pkg/podimpersonation/podimpersonation.go
Comment (tomleb) [inline pkg/podimpersonation/podimpersonation.go:584]: This should probably respect the[ image override that is done further down](https://github.com/rancher/steve/blob/dcea050d923061f427bb3f7e0a6a8f2a2d2992b9/pkg/podimpersonation/podimpersonation.go#L606-L609) in this function.

Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation.go]: Done

**Review by rohitsakala — APPROVED** [review]

File: pkg/podimpersonation/podimpersonation.go
Comment (tomleb) [inline pkg/podimpersonation/podimpersonation.go:576]: nit: Could add a comment to explain why we're doing this since it's not obvious at first

**Review by tomleb — APPROVED** [review]
One nit but otherwise LGTM.

Thanks for adding some tests.

**Comment by tomleb** [comment]
Needs one more approval from frameworks folks.

Also this PR will need someone from your team to QA it :+1:

**Comment by diogoasouza** [comment]
> Needs one more approval from frameworks folks.
> 
> Also this PR will need someone from your team to QA it 👍

Sure! I'll put it in our QA backlog after it's merged

File: pkg/podimpersonation/podimpersonation_test.go
Comment (JonCrowther) [inline pkg/podimpersonation/podimpersonation_test.go:46]:
```suggestion
				assert.Len(t, pod.Spec.Containers[0].VolumeMounts, len(p.Spec.Containers[0].VolumeMounts)+1, "expected kubeconfig volume to be mounted")
```

**Review by JonCrowther — COMMENTED** [review]
Everything looks good. Just found a minor typo

File: pkg/podimpersonation/podimpersonation_test.go
Comment (diogoasouza) [reply pkg/podimpersonation/podimpersonation_test.go]: Fixed it

**Review by JonCrowther — APPROVED** [review]

**Comment by diogoasouza** [comment]
I think it has the required approvals, can you merge it @tomleb ?

**Pull request merged**
